nixos/modules/services/networking/fedimintd.nix

```diff
@@ -0,0 +1,363 @@
+                ''
+                  secret=$(${pkgs.coreutils}/bin/head -n 1 "${cfg.bitcoin.rpc.secretFile}")
+                  prefix="''${FM_BITCOIN_RPC_URL%*@*}"  # Everything before the last '@'
+                  suffix="''${FM_BITCOIN_RPC_URL##*@}"  # Everything after the last '@'
+                  FM_BITCOIN_RPC_URL="''${prefix}:''${secret}@''${suffix}"
+                ''
+              else
+                ""
+            )
+            + ''
+              exec ${cfg.package}/bin/fedimintd
+            ''
+          );
+        in
+        {
+          description = "Fedimint Server";
+          documentation = [ "https://github.com/fedimint/fedimint/" ];
+          wantedBy = [ "multi-user.target" ];
+          environment = lib.mkMerge [
+            {
+              FM_BIND_P2P = "${cfg.p2p.bind}:${toString cfg.p2p.port}";
+              FM_BIND_API = "${cfg.api.bind}:${toString cfg.api.port}";
+              FM_P2P_URL = cfg.p2p.url;
+              FM_API_URL = cfg.api.url;
+              FM_DATA_DIR = cfg.dataDir;
+              FM_BITCOIN_NETWORK = cfg.bitcoin.network;
+              FM_BITCOIN_RPC_URL = cfg.bitcoin.rpc.url;
+              FM_BITCOIN_RPC_KIND = cfg.bitcoin.rpc.kind;
+            }
+            cfg.environment
+          ];
+          serviceConfig = {
+            User = cfg.user;
+            Group = cfg.group;
+
+            StateDirectory = "fedimintd-${fedimintdName}";
+            StateDirectoryMode = "0700";
+            ExecStart = startScript;
+
+            Restart = "always";
+            RestartSec = 10;
+            StartLimitBurst = 5;
+            UMask = "007";
+            LimitNOFILE = "100000";
+
+            LockPersonality = true;
+            MemoryDenyWriteExecute = "true";
+            NoNewPrivileges = "true";
+            PrivateDevices = "true";
+            PrivateMounts = true;
+            PrivateTmp = "true";
+            ProtectClock = true;
+            ProtectControlGroups = true;
+            ProtectHostname = true;
+            ProtectKernelLogs = true;
+            ProtectKernelModules = true;
+            ProtectKernelTunables = true;
+            ProtectSystem = "full";
+            RestrictAddressFamilies = [
+              "AF_INET"
+              "AF_INET6"
+            ];
+            RestrictNamespaces = true;
+            RestrictRealtime = true;
+            SystemCallArchitectures = "native";
+            SystemCallFilter = [
+              "@system-service"
+              "~@privileged"
+            ];
+          };
+        }
+      ))
+    ) eachFedimintd;
+
+    users.users = mapAttrs' (
+      fedimintdName: cfg:
+      (nameValuePair "fedimintd-${fedimintdName}" {
+        name = cfg.user;
+        group = cfg.group;
+        description = "Fedimint daemon user";
+        home = cfg.dataDir;
+        isSystemUser = true;
+      })
+    ) eachFedimintd;
+
+    users.groups = mapAttrs' (fedimintdName: cfg: (nameValuePair "${cfg.group}" { })) eachFedimintd;
```

Comment:
It would be great if we could make systemd's `DynamicUser` option work for this, so we don't need so many hardcoded system users with dynamic UID/GIDs. Could you try dropping the users, setting `systemd.services.<name>.serviceConfig.DynamicUser = true`, and see if it still works normally? Considering the UMask and the StateDirectory privileges, it seems likely to work